app/next-client-app/api/projects.ts

```diff
@@ -0,0 +1,58 @@
+"use server";
+
+import request from "@/lib/api/request";
+import { fetchAllPages } from "@/lib/api/utils";
+
+const fetchKeys = {
+  list: (filter?: string) => (filter ? `projects/?${filter}` : "projects/"),
+  projectsDataset: (dataset: string) => `projects/?dataset=${dataset}`,
+  project: (id: string) => `projects/${id}/`,
+};
+
+export async function getProjectsList(
+  filter?: string | undefined
+): Promise<PaginatedResponse<Project>> {
+  try {
+    return await request<Project>(fetchKeys.list(filter));
+  } catch (error) {
+    console.warn("Failed to fetch data.");
+    return { count: 0, next: null, previous: null, results: [] };
+  }
+}
+
+export async function getAllProjects(): Promise<Project[]> {
+  try {
+    // If there are more than 100 projects in Carrot, adjust the page size below
+    return await fetchAllPages<Project>(fetchKeys.list("page_size=100"));
```

Comment:
`page_size` should be an optional argument to this function instead of hardcoding it here.

Reply:
For some reason, the call will fail if the filter is empty. So for now, I will put an empty param here. Need more guidance here, actually

Reply:
I found out that the function `fetchAllPages` was designed to expect the passed URL to already have a param query character there (`?`). So if the param for `fetchAllPages<Project>(fetchKeys.list());` is empty, `fetchAllPages` will call the URL `api/projects/&p=1` (without `?`) which will fail.

So I think we can either modify the function `fetchAllPages` (and we then need to update other related functions/components using this) or put an empty param for it (like " " or "p") as a "fake" filter (then it will call the URL with `?`)